### DISCUSSION TITLE
뮤텍스와 세마포어의 차이를 설명해주세요.

### DISCUSSION THREAD

**Discussion by eatingbug** [Team-Psyduck/tech-sharing#66] [General]
-

**Comment by eatingbug** [comment]
뮤텍스는 Locking 메커니즘으로 락을 걸은 Thread 만이 임계 영역을 나갈때 락을 해제할 수 있다. 하지만 세마포어는 Signaling 메커니즘으로 락을 걸지 않은 쓰레드도 signal을 사용해 락을 해제할 수 있다. 
- 세마포어의 카운트를 1로 설정하면 뮤텍스처럼 활용할 수 있다.

출처 : https://mangkyu.tistory.com/104

**Comment by SionBackEnd** [comment]
뮤텍스와 세마포어의 차이점은 가질 수 있는 락의 개수 입니다. 뮤텍스는 락을 1개만 가질 수 있으며, 락을 가진 스레드만이 락을 해제할 수 있습니다. 하지만, 세마포어는 락을 1개 이상 가질 수 있으며 락을 가지고 있지 않은 스레드도 락을 해제할 수 있습니다. 
락을 1개만 가지고 있는 이진 세마포어를 사용하면 뮤텍스와 동일한 기능을 가질 수 있습니다.
